packages/apollo-server-core/src/plugin/usageReporting/plugin.ts

```diff
@@ -0,0 +1,720 @@
+         */
+        let endDone: boolean = false;
+        function didEnd(
+          requestContext:
+            | GraphQLRequestContextWillSendResponse<TContext>
+            // Our didEncounterErrors handler only calls this function if didResolveSource.
+            | (GraphQLRequestContextDidEncounterErrors<TContext> &
+                GraphQLRequestContextDidResolveSource<TContext>)
+            | GraphQLRequestContextDidResolveOperation<TContext>,
+        ) {
+          if (endDone) return;
+          endDone = true;
+          treeBuilder.stopTiming();
+
+          if (metrics.captureTraces === undefined) {
+            logger.warn(
+              'captureTrace is undefined at the end of the request. This is a bug in the Apollo Engine Reporting plugin.',
+            );
+          }
+
+          if (metrics.captureTraces === false) return;
+
+          treeBuilder.trace.fullQueryCacheHit = !!metrics.responseCacheHit;
+          treeBuilder.trace.forbiddenOperation = !!metrics.forbiddenOperation;
+          treeBuilder.trace.registeredOperation = !!metrics.registeredOperation;
+
+          // If operation resolution (parsing and validating the document followed
+          // by selecting the correct operation) resulted in the population of the
+          // `operationName`, we'll use that. (For anonymous operations,
+          // `requestContext.operationName` is null, which we represent here as
+          // the empty string.)
+          //
+          // If the user explicitly specified an `operationName` in their request
+          // but operation resolution failed (due to parse or validation errors or
+          // because there is no operation with that name in the document), we
+          // still put _that_ user-supplied `operationName` in the trace. This
+          // allows the error to be better understood in Graph Manager. (We are
+          // considering changing the behavior of `operationName` in these 3 error
+          // cases; https://github.com/apollographql/apollo-server/pull/3465)
+          const operationName =
+            requestContext.operationName ||
+            requestContext.request.operationName ||
+            '';
+
+          // If this was a federated operation and we're the gateway, add the query plan
+          // to the trace.
+          if (metrics.queryPlanTrace) {
+            treeBuilder.trace.queryPlan = metrics.queryPlanTrace;
+          }
+
+          // Intentionally un-awaited so as not to block the response.  Any
+          // errors will be logged, but will not manifest a user-facing error.
+          // The logger in this case is a request specific logger OR the logger
+          // defined by the plugin if that's unavailable.  The request-specific
+          // logger is preferred since this is very much coupled directly to a
+          // client-triggered action which might be more granularly tagged by
+          // logging implementations.
+          addTrace().catch(logger.error);
```

Comment:
We talked a bit on Slack. I'm going with setImmediate.

I'm also removing a similar line from `sendReport`, since the only place where `sendReport` is called from within a resolver is from `addTrace` itself.